src/User.cpp

```diff
@@ -852,6 +852,94 @@ bool CUser::IsHostAllowed(const CString& sHostMask) const {
     for (const CString& sHost : m_ssAllowedHosts) {
         if (sHostMask.WildCmp(sHost)) {
             return true;
+        } else {
+            // CIDR notation checker, e.g. "192.0.2.0/24" or "2001:db8::/32"
+
+            // Try to split the string into an IP and routing prefix
+            VCString vsSplitCIDR;
+            const int iSplitCIDRCount = sHost.Split("/", vsSplitCIDR, false);
+            const int iRoutingPrefix = vsSplitCIDR.back().ToInt();
+
+            if (iSplitCIDRCount != 2 || iRoutingPrefix < 0) continue;
+
+            // Convert each IP from a numeric string to an addrinfo
+            addrinfo aiHints;
+            memset(&aiHints, 0, sizeof(addrinfo));
+            aiHints.ai_flags = AI_NUMERICHOST;
+
+            addrinfo* aiHost;
+            int iIsHostValid =
+                getaddrinfo(sHostMask.c_str(), NULL, &aiHints, &aiHost);
+            if (iIsHostValid != 0) continue;
+
+            aiHints.ai_family = aiHost->ai_family;  // Host and range must be in
+                                                    // the same address family
+
+            addrinfo* aiRange;
+            int iIsRangeValid = getaddrinfo(vsSplitCIDR.front().c_str(), NULL,
+                                            &aiHints, &aiRange);
+            if (iIsRangeValid != 0) {
+                freeaddrinfo(aiHost);
+                continue;
+            }
+
+            // "/0" allows all IPv[4|6] addresses
+            if (iRoutingPrefix == 0) {
+                freeaddrinfo(aiHost);
+                freeaddrinfo(aiRange);
+                return true;
+            }
+
+            // If both IPs are valid and of the same type, make a bit field mask
+            // from the routing prefix, AND it to the host and range, and see if
+            // they match
+            bool bIsHostInRange = false;
+            if (aiHost->ai_family == AF_INET) {
+                const sockaddr_in* saHost = (sockaddr_in*)(aiHost->ai_addr);
+                const sockaddr_in* saRange = (sockaddr_in*)(aiRange->ai_addr);
+
+                // Make IPv4 bitmask
+                const in_addr_t inBitmask =
+                    htonl((~0u) << (32 - iRoutingPrefix));
```

Comment:
> Oversized Shift Amounts: Shifting a uint32_t by 32 or more bits is undefined. My guess is that this originated because the underlying shift operations on various CPUs do different things with this: [...]

http://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html